### ISSUE TITLE
Is the output shape calculation correct for 2d convolutions?

### ISSUE THREAD

**Issue by docrozza** [awslabs/djl#384]
I was recently looking at the code for Conv2d convolutions to check the output shapes for a model I'm building

https://github.com/awslabs/djl/blob/97fce77e51e61e649906d37a1722f2c9e0678c08/api/src/main/java/ai/djl/nn/convolutional/Conv2d.java#L41-L49

However, in the code this calculation isn't quite correct according to the documentation

https://github.com/awslabs/djl/blob/97fce77e51e61e649906d37a1722f2c9e0678c08/api/src/main/java/ai/djl/nn/convolutional/Convolution.java#L162-L168

The dilation and stride only ever use the 0-index rather than the i-index as used for the kernel and padding values. Whilst this is OK for square shapes, this could be an issue for other shapes.

The same also applies to the new deconvolution code but for stride and padding this time

https://github.com/awslabs/djl/blob/97fce77e51e61e649906d37a1722f2c9e0678c08/api/src/main/java/ai/djl/nn/convolutional/Deconvolution.java#L142-L146

I wanted to check that this was a real issue before looking at contributing a fix

**Comment by stu1130** [comment]
Yes you are right. Stride & Dilation should also change along with i variant. Would you like to submit a PR to fix it.
You can check reference calculation [here](https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html).

**Comment by docrozza** [comment]
sure - just checking out the contributing guidelines now

**Comment by docrozza** [comment]
The code for the Deconvolution also appears to be missing the dilation component too and not handling the kernel size correctly

See [here](https://pytorch.org/docs/stable/generated/torch.nn.ConvTranspose2d.html#torch.nn.ConvTranspose2d) for the Conv2dTranspose, i.e

outHeight = (inHeight − 1) × stride[0] − 2 × padding[0] + dilation[0] × (kernel_size[0] − 1) + output_padding[0] + 1
outWidth  = (inWidth  − 1) × stride[1] − 2 × padding[1] + dilation[1] × (kernel_size[1] − 1) + output_padding[1] + 1

What's the projects preference on this - fix it in this issue's PR or raise a separate one for just the Deconvolution logic?

**Comment by stu1130** [comment]
@docrozza You can fix it all in one PR
